src/NHibernate/Persister/Entity/AbstractEntityPersister.cs

```diff
@@ -3795,34 +3795,24 @@ public virtual void AfterReassociate(object entity, ISessionImplementor session)
 				return true;
 			}
 
-            // check the id unsaved-value
-            // We do this first so we don't have to hydrate the version property if the id property already gives us the info we need (NH-3505).
-            bool? result2 = entityMetamodel.IdentifierProperty.UnsavedValue.IsUnsaved(id);
-            if (result2.HasValue)
-            {
-                if (IdentifierGenerator is Assigned)
-                {
-                    // if using assigned identifier, we can only make assumptions
-                    // if the value is a known unsaved-value
```

Comment:
`if` block removed, just keeping the `else` body: `return result2;`.

(The whole block was space indented since NH-3505, thus the larger change.)

Reply:
What if we move this block after the version check?

Reply:
This?
```c#
// check the id unsaved-value
// We do this first so we don't have to hydrate the version property if the id property already gives us the info we need (NH-3505).
bool? result2 = entityMetamodel.IdentifierProperty.UnsavedValue.IsUnsaved(id);
if (result2.HasValue && !(IdentifierGenerator is Assigned))
{
	return result2;
}

// check the version unsaved-value, if appropriate
if (IsVersioned)
{
	object version = GetVersion(entity);
	bool? result = entityMetamodel.VersionProperty.UnsavedValue.IsUnsaved(version);
	if (result.HasValue)
	{
		return result;
	}
}

// if using assigned identifier, we can only make assumptions
// if the value is a known unsaved-value
if (result2 == true)
	return true;
```

That would still break `unsaved-value`. There is no reason at all for this block, excepted not understanding/knowing about `unsaved-value`. If we do not want any assumption to be done for assigned identifiers, we just need to have `unsaved-value="undefined"`, which is the default for assigned identifier. Specifying `unsaved-value="0"` in the mapping then complaining about NHibernate treating all entities with a non-zero value as already existing, is nonsense.